# Laboratorio 7 - Parte 2

## Integrantes

- Sergio Orellana 221122
- Rodrigo Mansilla 22611
- Carlos Valladares 221164

## Parte 2 - Teoría

### Pregunta 1
**Imagine que usted ha sido contratado para modelar el sistema de tráfico de una gran ciudad con el fin de evaluar el impacto de un nuevo sistema de metro. ¿Por qué un modelo puramente MBA, DES o DS sería, por sí solo, insuficiente para esta tarea? Describa cómo diseñaría usted un modelo híbrido para este problema, identificando qué componente del sistema sería modelado con cada paradigma y cómo interactuarían entre sí.**

Un modelo puramente MBA capturaría bien la heterogeneidad y las decisiones de los viajeros (cambio de ruta, hora de salida, adopción del metro), pero no representaría con precisión los cuellos de botella operativos ni los tiempos de servicio en estaciones y cruces. Un puramente DES modelaría con detalle las colas, recursos y flujos en andenes, tornos y semáforos, pero carecería de comportamiento adaptativo y de interacción social a nivel individual. Un puramente DS ofrecería una vista macro de stocks y flujos (adopción del metro, congestión promedio), pero sería demasiado agregado para políticas que dependen de microinteracciones y de la red vial concreta.  
Diseño un híbrido tri-paradigma:  
• **MBA (micro):** agentes personas y vehículos con preferencias, valor del tiempo, aprendizaje y respuesta a información en tiempo real; operadores que ajustan oferta/frecuencias.  
• **DES (operativo):** estaciones y nodos (torniquetes, andenes, escaleras, transbordos) y cruces/semaforización como procesos con recursos y colas; cálculo de tiempos de espera, utilización y throughput por franja.  
• **DS (macro):** bucles de realimentación para adopción del metro y demanda total, congestión promedio y elasticidades que devuelven tasas y niveles a MBA y DES.  
**Acoplo:** MBA genera llegadas por estación/hora hacia el DES; el DES devuelve tiempos de espera y fiabilidad que modifican la utilidad percibida en MBA; el DS lee agregados (congestión, puntualidad, saturación) y actualiza tendencias de demanda que condicionan la población simulada (MBA) y la capacidad operativa (DES).

### Pregunta 2
**En el "Patrón 1: MBA → DES", la transición de un agente a una entidad debe ser fluida. Describa a nivel conceptual y técnico (¿qué información específica se debe transferir?) cómo se realizaría esta transición en una herramienta como AnyLogic, o de forma más abstracta, en un marco de trabajo en Python que combinara las bibliotecas Mesa (para MBA) y SimPy (para DES).**

Conceptualmente, cuando un agente “entra al proceso” (por ejemplo, llega a la estación), debo materializarlo como entidad de DES con los atributos mínimos para cola y servicio.  
Datos que transfiero: id del agente y hora de llegada (para medir espera), tipo de servicio requerido, prioridad, distribución/estimación de tiempo de servicio, ubicación/estación y ruta de proceso, además de estado relevante (p. ej., tolerancia a esperar o probabilidad de abandono).  
En un stack Mesa+SimPy, el agente crea un mensaje/objeto con esos campos y lo encola en una Store o pasa a un proceso (`env.process(servicio(entidad))`) que hace seize/hold/release sobre `Resource` o `PriorityResource`. Al terminar, el DES emite un evento de salida con tiempos de espera y servicio y el resultado (éxito/abandono) para que el agente en Mesa actualice su estado y decida su siguiente acción. En AnyLogic, mapearía propiedades del agente a atributos de la entidad al entrar a la cadena de bloques de proceso y sincronizaría de vuelta al abandonar el flujo.

### Pregunta 3
**El "Patrón 3: DS dentro de un Agente" es un claro ejemplo de modelado multi-escala. Proponga usted otro escenario (fuera del ámbito biológico), por ejemplo, en economía o sociología, donde este patrón sería de gran utilidad. ¿Qué representaría el agente (MBA) y qué dinámica capturaría su modelo de Dinámica de Sistemas (DS) interno?**

Planteo un mercado minorista con agentes‑empresa (tiendas). Cada tienda es un agente que compite por demanda local y decide precios/promociones, pero dentro lleva un modelo DS de inventario–pedidos–backlog (stocks: inventario disponible, pedidos en tránsito y cartera; flujos: ventas, reposición y devoluciones). Ese DS interno capta retrasos y bucles (stock bajo → más pedidos → posible sobrestock al arribar) que explican roturas de stock y efecto látigo. El MBA coordina la competencia espacial y la interacción con clientes heterogéneos; el DS interno gobierna la política de reposición y la dinámica operativa de cada tienda.

### Pregunta 4
**La combinación de paradigmas de modelado presenta desafíos significativos. Desde su perspectiva, ¿cuáles cree usted que son los dos mayores desafíos al construir y validar un modelo híbrido? Considere aspectos como la calibración de parámetros entre los diferentes componentes, la complejidad del software y la validación de los resultados emergentes.**

Primero, la calibración y consistencia entre escalas/paradigmas: debo alinear tasas y tiempos del DES (distribuciones de servicio y llegada), reglas del MBA (preferencias, elasticidades, aprendizaje) y parámetros del DS (ganancias de bucle, retardos) para describir la misma realidad, traduciendo métricas observables (esperas, utilización) en utilidades/comportamientos y evitando doble contabilidad de efectos. Segundo, la complejidad e inferencia sobre fenómenos emergentes: debo validar no solo series agregadas, sino patrones emergentes (distribución de viajes, formación y disipación de colas, adopción del metro), con sensibilidad y escenarios para garantizar robustez, además de verificación y pruebas de rendimiento del software.



# Referencias

- GeeksforGeeks. (2020, November 19). Basics of Discrete Event Simulation using SimPy. GeeksforGeeks. https://www.geeksforgeeks.org/python/basics-of-discrete-event-simulation-using-simpy/ 
- GeeksforGeeks. (2024, May 27). Queuing theory. GeeksforGeeks. https://www.geeksforgeeks.org/maths/queuing-theory/ 
- GeeksforGeeks. (2025a, July 23). Agent based modelling. GeeksforGeeks. https://www.geeksforgeeks.org/artificial-intelligence/agent-based-modelling/ 
- GeeksforGeeks. (2025b, July 23). How to do Mathematical Modeling in Python? GeeksforGeeks. https://www.geeksforgeeks.org/python/how-to-do-mathematical-modeling-in-python/ 
- GeeksforGeeks. (2025c, July 23). What is Systems Thinking? | Working, Benefits and Limitations. GeeksforGeeks. https://www.geeksforgeeks.org/software-engineering/what-is-systems-thinking-working-benefits-and-limitations/ 
- GeeksforGeeks. (2025d, August 14). Agents in AI. GeeksforGeeks. https://www.geeksforgeeks.org/artificial-intelligence/agents-artificial-intelligence/

---

## Práctica: Modelo Híbrido (MBA + DS + DES)

### Descripción del Modelo

Este modelo integra tres paradigmas:
- **MBA (Agent-Based Model)**: Agentes que se mueven e interactúan en el espacio
- **DS (System Dynamics)**: Cada agente tiene un stock de energía con flujos de entrada/salida
- **DES (Discrete Event Simulation)**: Estación de recarga como recurso limitado con colas

### 1. Importar Bibliotecas

In [1]:
import simpy
import random
import numpy as np
import matplotlib.pyplot as plt
from collections import defaultdict

### 2. Definir Parámetros del Modelo

In [2]:
NUM_AGENTES = 20
ANCHO_MUNDO = 100.0
ALTO_MUNDO = 100.0

TASA_RECUPERACION_NATURAL = 0.1     
GASTO_POR_MOVIMIENTO = 0.05          
EFECTO_INTERACCION = 0.2             

NUM_PUESTOS_RECARGA = 5             
ENERGIA_CRITICA = 2.0               
TIEMPO_RECARGA = 10.0               
ENERGIA_MAXIMA = 10.0               
TIEMPO_SIMULACION = 200.0            
DT = 1.0                             

RADIO_INTERACCION = 5.0

### 3. Clase Agente (Integración MBA + DS + DES)

Esta clase integra los tres paradigmas:
- **Atributos MBA**: posición, velocidad
- **Atributos DS**: energía (stock), tasas de flujo
- **Métodos DES**: proceso de recarga con recursos limitados

In [ ]:
class Agente:
    def __init__(self, id_agente, posicion_inicial, env, estacion_recarga):

        self.id = id_agente
        
        self.posicion = np.array(posicion_inicial, dtype=float)
        angulo = random.uniform(0, 2 * np.pi)
        velocidad_magnitud = random.uniform(0.5, 2.0)
        self.velocidad = np.array([
            velocidad_magnitud * np.cos(angulo),
            velocidad_magnitud * np.sin(angulo)
        ])
        
        self.energia = ENERGIA_MAXIMA  

        self.env = env
        self.estacion_recarga = estacion_recarga
        self.en_recarga = False  
        
        self.veces_recargado = 0
        self.tiempo_total_esperando = 0.0
    
    def proceso_recarga(self):
        
        if self.en_recarga:
            return  
        
        self.en_recarga = True
        tiempo_inicio_espera = self.env.now
        
        print(f"  [DES] Agente {self.id} solicita recarga en t={self.env.now:.1f} (energía={self.energia:.2f})")
        
        with self.estacion_recarga.request() as req:
            yield req  
            
            tiempo_espera = self.env.now - tiempo_inicio_espera
            self.tiempo_total_esperando += tiempo_espera
            
            print(f"  [DES] Agente {self.id} inicia recarga en t={self.env.now:.1f} (esperó {tiempo_espera:.1f} min)")
            
            yield self.env.timeout(TIEMPO_RECARGA)
            
            self.energia = ENERGIA_MAXIMA
            self.veces_recargado += 1
            
            print(f"  [DES] Agente {self.id} termina recarga en t={self.env.now:.1f} (energía={self.energia:.2f})")
        
        self.en_recarga = False
    
    def actualizar_estado(self, efecto_interaccion_neto, dt):
      
        recuperacion = TASA_RECUPERACION_NATURAL * dt
        
        velocidad_magnitud = np.linalg.norm(self.velocidad)
        gasto_movimiento = GASTO_POR_MOVIMIENTO * velocidad_magnitud * dt
        
        efecto_interacciones = efecto_interaccion_neto * dt
        
        self.energia += recuperacion - gasto_movimiento + efecto_interacciones
        
        self.energia = min(self.energia, ENERGIA_MAXIMA)
        

        self.posicion += self.velocidad * dt
        
        if self.posicion[0] <= 0 or self.posicion[0] >= ANCHO_MUNDO:
            self.velocidad[0] = -self.velocidad[0]
            self.posicion[0] = np.clip(self.posicion[0], 0, ANCHO_MUNDO)
        
        if self.posicion[1] <= 0 or self.posicion[1] >= ALTO_MUNDO:
            self.velocidad[1] = -self.velocidad[1]
            self.posicion[1] = np.clip(self.posicion[1], 0, ALTO_MUNDO)
        
        if self.energia < ENERGIA_CRITICA and not self.en_recarga:
            self.env.process(self.proceso_recarga())

### 4. Función Principal de Simulación

Esta función orquesta el bucle principal que integra:
1. **MBA**: Detección de interacciones entre agentes cercanos
2. **DS**: Actualización de stocks de energía
3. **DES**: Gestión de eventos de recarga
4. **Recolección de datos** para análisis posterior

In [ ]:
def ejecutar_simulacion(env, agentes, tiempo_simulacion):

    historial_energia = []
    historial_agentes_en_cola = []
    historial_tiempo = []
    
 
    print(f"Agentes: {len(agentes)}")
    print(f"Puestos de recarga: {NUM_PUESTOS_RECARGA}")
    print(f"Duración: {tiempo_simulacion} unidades de tiempo")

    
    while env.now < tiempo_simulacion:
        tiempo_actual = env.now
        
        efectos_interaccion = defaultdict(float)
        
        for i, agente_i in enumerate(agentes):
            for j, agente_j in enumerate(agentes):
                if i >= j:  
                    continue
                
                distancia = np.linalg.norm(agente_i.posicion - agente_j.posicion)
                
                if distancia < RADIO_INTERACCION:
                    efecto = random.choice([-1, 1]) * EFECTO_INTERACCION
                    
                    efectos_interaccion[agente_i.id] += efecto
                    efectos_interaccion[agente_j.id] += efecto
        
        for agente in agentes:
            efecto_neto = efectos_interaccion.get(agente.id, 0.0)
            agente.actualizar_estado(efecto_neto, DT)
        

        energia_promedio = np.mean([a.energia for a in agentes])
        historial_energia.append(energia_promedio)
        
        
        agentes_en_cola = sum(1 for a in agentes if a.en_recarga)
        historial_agentes_en_cola.append(agentes_en_cola)
        
        historial_tiempo.append(tiempo_actual)
        
        if int(tiempo_actual) % 20 == 0 and tiempo_actual > 0:
            energia_min = min(a.energia for a in agentes)
            energia_max = max(a.energia for a in agentes)
            print(f"t={tiempo_actual:.0f}: Energía promedio={energia_promedio:.2f}, "
                  f"Min={energia_min:.2f}, Max={energia_max:.2f}, "
                  f"En recarga={agentes_en_cola}")
        
        yield env.timeout(DT)
    

    
    return {
        'tiempo': historial_tiempo,
        'energia_promedio': historial_energia,
        'agentes_en_cola': historial_agentes_en_cola,
        'agentes': agentes
    }


### 5. Configuración y Ejecución - Experimento 1

**Configuración**: NUM_PUESTOS_RECARGA = 5 (capacidad generosa)

In [6]:
NUM_PUESTOS_RECARGA = 5

env = simpy.Environment()

estacion_recarga = simpy.Resource(env, capacity=NUM_PUESTOS_RECARGA)

agentes = []
for i in range(NUM_AGENTES):
    pos_inicial = [
        random.uniform(0, ANCHO_MUNDO),
        random.uniform(0, ALTO_MUNDO)
    ]
    agente = Agente(i, pos_inicial, env, estacion_recarga)
    agentes.append(agente)

resultados_exp1 = {}

def simulacion_con_retorno():
    resultado = yield env.process(ejecutar_simulacion(env, agentes, TIEMPO_SIMULACION))
    global resultados_exp1
    resultados_exp1 = resultado

env.process(simulacion_con_retorno())
env.run()

Agentes: 20
Puestos de recarga: 5
Duración: 200.0 unidades de tiempo
t=20: Energía promedio=9.86, Min=9.03, Max=10.00, En recarga=0
t=40: Energía promedio=9.69, Min=7.59, Max=10.00, En recarga=0
t=60: Energía promedio=9.70, Min=7.55, Max=10.00, En recarga=0
t=80: Energía promedio=9.72, Min=7.97, Max=10.00, En recarga=0
t=100: Energía promedio=9.71, Min=8.08, Max=10.00, En recarga=0
t=120: Energía promedio=9.70, Min=6.81, Max=10.00, En recarga=0
t=140: Energía promedio=9.71, Min=6.96, Max=10.00, En recarga=0
t=160: Energía promedio=9.74, Min=7.12, Max=10.00, En recarga=0
t=180: Energía promedio=9.74, Min=8.07, Max=10.00, En recarga=0


### 6. Análisis de Resultados - Experimento 1

In [8]:
if resultados_exp1:

    
    total_recargas = sum(a.veces_recargado for a in agentes)
    agentes_que_recargaron = sum(1 for a in agentes if a.veces_recargado > 0)
    
    print(f"\n Uso de la Estación de Recarga:")
    print(f"   • Total de recargas realizadas: {total_recargas}")
    print(f"   • Agentes que usaron la estación: {agentes_que_recargaron}/{NUM_AGENTES}")
    print(f"   • Promedio de recargas por agente: {total_recargas/NUM_AGENTES:.2f}")
    
    if agentes_que_recargaron > 0:
        tiempo_espera_promedio = sum(a.tiempo_total_esperando for a in agentes) / agentes_que_recargaron
        print(f"   • Tiempo promedio de espera en cola: {tiempo_espera_promedio:.2f} unidades")
    
    # Análisis de energía
    energia_final_promedio = np.mean([a.energia for a in agentes])
    energia_final_min = min(a.energia for a in agentes)
    energia_final_max = max(a.energia for a in agentes)
    
    print(f"\n Estado Final de Energía:")
    print(f"   • Energía promedio: {energia_final_promedio:.2f}")
    print(f"   • Energía mínima: {energia_final_min:.2f}")
    print(f"   • Energía máxima: {energia_final_max:.2f}")



 Uso de la Estación de Recarga:
   • Total de recargas realizadas: 0
   • Agentes que usaron la estación: 0/20
   • Promedio de recargas por agente: 0.00

 Estado Final de Energía:
   • Energía promedio: 9.75
   • Energía mínima: 8.22
   • Energía máxima: 10.00



### 8. Experimento 2: Cuello de Botella

**Configuración**: NUM_PUESTOS_RECARGA = 1 (capacidad limitada)

In [16]:
NUM_PUESTOS_RECARGA = 1 

env2 = simpy.Environment()

estacion_recarga2 = simpy.Resource(env2, capacity=NUM_PUESTOS_RECARGA)

agentes2 = []
for i in range(NUM_AGENTES):
    pos_inicial = [
        random.uniform(0, ANCHO_MUNDO),
        random.uniform(0, ALTO_MUNDO)
    ]
    agente = Agente(i, pos_inicial, env2, estacion_recarga2)
    agentes2.append(agente)

resultados_exp2 = {}

def simulacion_con_retorno2():
    resultado = yield env2.process(ejecutar_simulacion(env2, agentes2, TIEMPO_SIMULACION))
    global resultados_exp2
    resultados_exp2 = resultado

env2.process(simulacion_con_retorno2())
env2.run()

Agentes: 20
Puestos de recarga: 1
Duración: 200.0 unidades de tiempo
t=20: Energía promedio=9.88, Min=9.07, Max=10.00, En recarga=0
t=40: Energía promedio=9.95, Min=9.16, Max=10.00, En recarga=0
t=60: Energía promedio=9.97, Min=9.46, Max=10.00, En recarga=0
t=80: Energía promedio=9.87, Min=9.33, Max=10.00, En recarga=0
t=100: Energía promedio=9.92, Min=9.51, Max=10.00, En recarga=0
t=120: Energía promedio=9.95, Min=9.47, Max=10.00, En recarga=0
t=140: Energía promedio=9.95, Min=9.60, Max=10.00, En recarga=0
t=160: Energía promedio=9.90, Min=9.04, Max=10.00, En recarga=0
t=180: Energía promedio=9.93, Min=9.54, Max=10.00, En recarga=0


### 9. Análisis de Resultados - Experimento 2

In [20]:
if resultados_exp2:
    total_recargas2 = sum(a.veces_recargado for a in agentes2)
    agentes_que_recargaron2 = sum(1 for a in agentes2 if a.veces_recargado > 0)
    
    print(f"\n Uso de la Estación de Recarga:")
    print(f"   • Total de recargas realizadas: {total_recargas2}")
    print(f"   • Agentes que usaron la estación: {agentes_que_recargaron2}/{NUM_AGENTES}")
    print(f"   • Promedio de recargas por agente: {total_recargas2/NUM_AGENTES:.2f}")
    
    energia_final_promedio2 = np.mean([a.energia for a in agentes2])
    energia_final_min2 = min(a.energia for a in agentes2)
    energia_final_max2 = max(a.energia for a in agentes2)
    
    print(f"Estado Final de Energía:")
    print(f"   • Energía promedio: {energia_final_promedio2:.2f}")
    print(f"   • Energía mínima: {energia_final_min2:.2f}")
    print(f"   • Energía máxima: {energia_final_max2:.2f}")
    
    agentes_criticos = sum(1 for a in agentes2 if a.energia < ENERGIA_CRITICA * 1.5)
    print(f"   • Agentes con energía baja (<{ENERGIA_CRITICA * 1.5}): {agentes_criticos}")
    

else:
    print("No se pudieron obtener resultados")


 Uso de la Estación de Recarga:
   • Total de recargas realizadas: 0
   • Agentes que usaron la estación: 0/20
   • Promedio de recargas por agente: 0.00
Estado Final de Energía:
   • Energía promedio: 9.90
   • Energía mínima: 9.43
   • Energía máxima: 10.00
   • Agentes con energía baja (<3.0): 0



### Preguntas de Análisis

1. **¿Cuántos agentes, en promedio, utilizan la estación durante la simulación con 5 puestos?**
   - 0 agentes en promedio. La utilización de la estación fue **0%** sin recargas. Dado que con parámetros actuales los agentes mantienen energías cercanas al máximo, aun con 5 puestos no se activa el proceso DES.

2. **¿Cómo impacta reducir a 1 puesto en la dinámica general de energía?**
   - Sin impacto observable bajo estos parámetros. Con 1 puesto se registró:
   - **Total de recargas:** 0  
   - **Energía promedio final:** 9.90 (min 9.43, máx 10.00)  
   - **Agentes con energía < 3.0:** 0  
   La población se mantiene en un régimen no restrictivo: la energía promedio permanece alta y estable.

3. **¿Qué efecto tiene el cuello de botella DES sobre el comportamiento agregado MBA+DS?**
   - **Ninguno** en estas corridas. El supuesto cuello de botella **no se activa** (no hay demanda de recarga), por lo que la energía promedio se mantiene entre **9.87–9.97** a lo largo del tiempo y no se observan colas ni degradación de la dinámica agregada.


4. **Describa el bucle de retroalimentación entre DES y MBA+DS**
   - En el esquema previsto, baja energía → demanda de recarga (DES) → colas si la capacidad es insuficiente → mayor tiempo sin actividad útil → efectos sobre interacción y movimiento (MBA) → cambios en los flujos de energía (DS).  En estas simulaciones, el recuperación natural domina y el gasto por movimiento/interacción es bajo; por ello no se cruza el umbral crítico y el bucle DES↔MBA+DS permanece inactivo.


---

### Uso de Gen AI - Documentación

#### Prompt Utilizado:
```
Ayúdame a implementar un modelo híbrido en Python con SimPy que integre MBA, DS y DES. 
Necesito:

1. Una clase Agente que tenga:
   - Atributos MBA: posición, velocidad, movimiento en 2D
   - Atributos DS: energía como stock con flujos (recuperación, gasto por movimiento)
   - Métodos DES: proceso de recarga usando recursos SimPy con colas

2. Función principal que:
   - Detecte interacciones entre agentes cercanos (MBA)
   - Actualice energía con integración de Euler (DS)
   - Gestione eventos de recarga cuando energía < umbral crítico (DS→DES)
   - Recolecte datos para análisis temporal

3. Explica cómo se integran los tres paradigmas y dónde ocurren las transiciones

Quiero entender la arquitectura del modelo híbrido, no solo el código.
```

#### ¿Por qué funcionó este prompt?

1. **Especifica los tres paradigmas**: Dejé claro que necesito MBA, DS y DES integrados
2. **Estructura clara**: Separé por componentes (clase, función, integración)
3. **Énfasis en transiciones**: Pedí explícitamente cómo se conectan los paradigmas (DS→DES, MBA→DS)
4. **Contexto técnico**: Mencioné SimPy, integración de Euler, recursos con colas
5. **Objetivo de aprendizaje**: Pedí entender la arquitectura, no solo copiar código

